### PR TITLE
BottleLoader: Use the formula stored in the bottle

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

When installing a bottle from the local file system or on http, use the formula stored in the bottle.

### CODE REVIEW

File: Library/Homebrew/formulary.rb [hunk 1]
```diff
@@ -122,9 +122,22 @@ def initialize(bottle_name)
...
     end
 
+    def load_file
+      formula_path = HOMEBREW_CACHE_FORMULA/"#{name}.rb"
+      formula_path.dirname.mkpath
+      FileUtils.rm_f formula_path
+      bottle_version = Utils::Bottles.resolve_version @bottle_filename
+      system "tar", "-xf", @bottle_filename, "-C", HOMEBREW_CACHE_FORMULA,
+        "--strip-components", "3", "#{name}/#{bottle_version}/.brew/#{name}.rb",
+        err: :close
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: It would be good to avoid creating/deleting/writing files in the formulary loader. How about using something like `tar -O` to read it from `stdout`?

Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Good suggestion, Mike. Done.

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -123,7 +123,16 @@ def initialize(bottle_name)
     end
 
     def get_formula(spec, alias_path: nil)
-      formula = super
+      bottle_version = Utils::Bottles.resolve_version @bottle_filename
+      path = "#{name}/#{bottle_version}/.brew/#{name}.rb"
+      contents = Utils.popen_read "tar", "-xO", "-f", @bottle_filename,
+        "--strip-components", "3", path, err: :close
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: I don't want `err: :close` to be going throughout the codebase when they are no-ops for most users and when, in this case, the error should never be hit. I'm tempted to say rather than falling back this should raise an exception instead. Thoughts?

File: Library/Homebrew/formulary.rb [hunk 3]
```diff
@@ -123,7 +123,16 @@ def initialize(bottle_name)
...
+      formula = if $CHILD_STATUS.success?
+        Formulary.from_contents name, HOMEBREW_CELLAR/path, contents, spec
+      else
+        super
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: If this stays as-is: worth printing a `Warning` message in this case?

File: Library/Homebrew/formulary.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: An error is expected when used with a bottle that does not contain `.brew/#{name.rb}`, as for bottles older than about a year so. For these bottles without an internal formula, the behaviour of `brew install` remains unchanged.

File: Library/Homebrew/formulary.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: I've added a warning.

File: Library/Homebrew/formulary.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Does it make more sense to change the warning to an error and get rid of the old code path?

Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Agreed. I'll change it.

File: Library/Homebrew/formulary.rb [hunk 4]
```diff
@@ -123,7 +123,17 @@ def initialize(bottle_name)
...
     def get_formula(spec, alias_path: nil)
-      formula = super
+      bottle_version = Utils::Bottles.resolve_version @bottle_filename
+      path = "#{name}/#{bottle_version}/.brew/#{name}.rb"
+      contents = Utils.popen_read "tar", "-xO", "-f", @bottle_filename,
+        "--strip-components", "3", path, err: :close
+      formula = if $CHILD_STATUS.success?
+        Formulary.from_contents name, HOMEBREW_CELLAR/path, contents, spec
+      else
+        opoo "This bottle does not contain a formula: #{path}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: I think this should probably just fail rather than have inconsistent behaviour. If you're loading bottles more than a year old you're going to have other problems.

Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Agreed. Done.

File: Library/Homebrew/formulary.rb [hunk 5]
```diff
@@ -122,14 +122,14 @@ def initialize(bottle_name)
       super name, Formulary.path(full_name)
     end
 
-    def get_formula(spec, alias_path: nil)
-      formula = super
+    def get_formula(spec, alias_path: nil) # rubocop:disable Lint/UnusedMethodArgument
```
Comment (sjackman) [inline Library/Homebrew/formulary.rb]: Is `# rubocop:disable Lint/UnusedMethodArgument` the best way to avoid this `brew style` error?
```
W:125: 27: Unused method argument - alias_path.
```
See https://github.com/bbatsov/rubocop/issues/3130

Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Would you prefer…
```ruby
raise "Unexpected non-nil parameter alias_path" unless alias_path.nil?
```

**Comment by sjackman** [comment]
I have a failed test to address. I'll get back to you with that later today.
```
  1) Formulary::factory returns a Formula when given a bottle
     Failure/Error: formula = subject.factory(bottle)
     
     BottleFormulaUnavailableError:
       This bottle does not contain the formula file:
         testball_bottle/0.1/.brew/testball_bottle.rb
```

File: Library/Homebrew/formulary.rb (hunk 5 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formulary.rb]: I think `def get_formula(spec, **)` should work.

File: Library/Homebrew/exceptions.rb [hunk 6]
```diff
@@ -555,12 +555,11 @@ def initialize(resource)
...
 
-class BottleVersionMismatchError < RuntimeError
-  def initialize(bottle_file, bottle_version, formula, formula_version)
+class BottleFormulaUnavailableError < RuntimeError
+  def initialize(formula_path)
     super <<-EOS.undent
-      Bottle version mismatch
-      Bottle: #{bottle_file} (#{bottle_version})
-      Formula: #{formula.full_name} (#{formula_version})
+      This bottle does not contain the formula file:
```
Comment (MikeMcQuaid) [inline Library/Homebrew/exceptions.rb]: How about outputting the bottle file and expected formula path within it?

File: Library/Homebrew/formulary.rb [hunk 7]
```diff
@@ -122,14 +122,14 @@ def initialize(bottle_name)
...
+      bottle_version = Utils::Bottles.resolve_version @bottle_filename
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: Can this fail?

File: Library/Homebrew/formulary.rb [hunk 8]
```diff
@@ -122,14 +122,14 @@ def initialize(bottle_name)
...
+      formula_path = "#{name}/#{bottle_version}/.brew/#{name}.rb"
+      contents = Utils.popen_read "tar", "-xO", "-f", @bottle_filename, "--strip-components=3", formula_path
+      raise BottleFormulaUnavailableError, formula_path unless $CHILD_STATUS.success?
+      formula = Formulary.from_contents name, HOMEBREW_CELLAR/formula_path, contents, spec
+      formula.bottle_specification.sha256 @bottle_filename.sha256 => Utils::Bottles.tag
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: What happens without this, out of interest? I'm 👍 on it, just curious.

File: Library/Homebrew/formulary.rb (hunk 5 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Yes, that works, but it does alter the behavior of the method for style, in that `get_formula(spec, monkey: :banana)` no longer fails. https://github.com/bbatsov/rubocop/issues/3130#issuecomment-274852535
My minor style preference is `raise … unless nil?`. I'm fine with `**` though.

File: Library/Homebrew/formulary.rb (hunk 7 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: If the bottle does not contain a file matching the regex `.+/.+/INSTALL_RECEIPT.json`, then yes, it could fail. It currently throws an exception:
```
Error: undefined method `split' for nil:NilClass
/Users/sjackman/.homebrew/Library/Homebrew/utils/bottles.rb:40:in `resolve_formula_names'
```

Comment (sjackman) [reply Library/Homebrew/formulary.rb]: I've added
```diff
--- a/Library/Homebrew/utils/bottles.rb
+++ b/Library/Homebrew/utils/bottles.rb
@@ -29,9 +29,11 @@ module Utils
       end
 
       def receipt_path(bottle_file)
-        Utils.popen_read("tar", "-tzf", bottle_file).lines.map(&:chomp).find do |line|
+        path = Utils.popen_read("tar", "-tzf", bottle_file).lines.map(&:chomp).find do |line|
           line =~ %r{.+/.+/INSTALL_RECEIPT.json}
         end
+        raise "This bottle does not contain the file INSTALL_RECEIPT.json: #{bottle_file}" if path.nil?
+        path
       end
 
       def resolve_formula_names(bottle_file)
```

File: Library/Homebrew/formulary.rb (hunk 8 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Because the formula stored in the bottle does not have a `bottle` block, the `bottle` method returns `nil`, and causes the error. In principle that error would have also been possible before if the formula in the repo did not contain a bottle block, which is uncommon.
```
==> Pouring the cached bottle
Error: undefined method `compatible_cellar?' for nil:NilClass
/Users/sjackman/.homebrew/Library/Homebrew/formula_installer.rb:104:in `pour_bottle?'
```
I'll add a commit to address this issue.

File: Library/Homebrew/exceptions.rb (hunk 6 above)
Comment (sjackman) [reply Library/Homebrew/exceptions.rb]: Done.

File: Library/Homebrew/formulary.rb (hunk 5 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Done.

File: Library/Homebrew/formulary.rb (hunk 8 above)
Comment (sjackman) [reply Library/Homebrew/formulary.rb]: Done.

File: Library/Homebrew/formula_installer.rb [hunk 9]
```diff
@@ -101,7 +101,7 @@ def pour_bottle?(install_bottle_options = { warn: false })
       return false
     end
 
-    unless bottle.compatible_cellar?
+    if bottle && !bottle.compatible_cellar?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: `unless bottle&.compatible_cellar?`

File: Library/Homebrew/utils/bottles.rb [hunk 10]
```diff
@@ -29,9 +29,11 @@ def native_regex
       end
 
       def receipt_path(bottle_file)
-        Utils.popen_read("tar", "-tzf", bottle_file).lines.map(&:chomp).find do |line|
+        path = Utils.popen_read("tar", "-tzf", bottle_file).lines.map(&:chomp).find do |line|
           line =~ %r{.+/.+/INSTALL_RECEIPT.json}
         end
+        raise "This bottle does not contain the file INSTALL_RECEIPT.json: #{bottle_file}" if path.nil?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/bottles.rb]: `unless path`

File: Library/Homebrew/formula_installer.rb (hunk 9 above)
Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: I didn't know about the `&.` operator! Fantastic!

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_installer.rb]: Will be enforced by `brew style` from https://github.com/Homebrew/homebrew-portable/pull/50. Wasn't supported in Ruby 2.0.

Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: `formula.bottle` is `nil` when `bottle.compatible_cellar?` is false.
Using `formula.bottle_specification.compatible_cellar?` rather
than `formula.bottle.compatible_cellar?` avoids this problem

File: Library/Homebrew/utils/bottles.rb (hunk 10 above)
Comment (sjackman) [reply Library/Homebrew/utils/bottles.rb]: Done.

**Comment by sjackman** [comment]
I found one odd use case: `brew install -s local_bottle.tar.gz` failed when copying the formula to `.brew/NAME.rb`, because the formula is inside the bottle and not stored anywhere on disk. I've fixed this issue with commit 731b1b1e9c36d59ce48c05799a3ddc172d916dd8.

**Comment by sjackman** [comment]
Do you think it would be useful to add a `contents` method to the `Formula` class, that returns the Ruby code of the formula?

**Comment by MikeMcQuaid** [comment]
@sjackman If there's 3 or more locations it would be used: yep, otherwise: nope.

**Comment by ilovezfs** [comment]
Masking the formula that is in the git repository in favor of one buried in a binary and, in particular, allowing that to override the checksum, seems like a security risk to me.

**Comment by MikeMcQuaid** [comment]
> Masking the formula that is in the git repository in favor of one buried in a binary and, in particular, allowing that to override the checksum, seems like a security risk to me.

I think the checksum should be able to be set if it's unset if that's needed for internal code but it shouldn't be able to be overridden, I agree. I wonder if it's worth reading the version from the both formulae and using the one in Git if the version/revision/`pkg_version` match (as otherwise we are unlikely to have old bottles remain working after e.g. major `post_install` changes). Alternatively, the best middle ground might be only using the `post_install` from the bottled formula as that's the only bit that affects bottles at all.

**Comment by ilovezfs** [comment]
If we patch a critical CVE that would still let someone blithely `brew install ./badversion` since the versions wouldn't be the same.

**Comment by sjackman** [comment]
Sure. I can do that.

**Comment by sjackman** [comment]
My most recent commit displays a warning when downgrading a formula.

**Comment by sjackman** [comment]
> I think the checksum should be able to be set if it's unset if that's needed for internal code

Adding the `sha256` to the bottle spec is no longer necessary thanks to fixing up the logic in `pour_bottle?` not to freak out when `formula.bottle` is nil. I've removed the line that sets the `sha256` of the bottle.

File: Library/Homebrew/formula_installer.rb [hunk 11]
```diff
@@ -101,6 +100,7 @@ def pour_bottle?(install_bottle_options = { warn: false })
       return false
     end
 
+    bottle = formula.bottle_specification
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Why is this needed below, out of interest, and `bottle_specification` rather than `bottle`?

File: Library/Homebrew/formula_installer.rb [hunk 12]
```diff
@@ -235,6 +235,10 @@ def install
       end
       raise CannotInstallFormulaError, message
     end
+    if formula.opt_prefix.directory? &&
+       formula.pkg_version < (v = Keg.new(formula.opt_prefix.resolved_path).version)
+      opoo "Downgrading #{formula.name} from #{v} to #{formula.pkg_version}."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Sorry, rather than downgrading here would be good to check the tab in the bottle and see if the formula file exists and that one has a newer version.

File: Library/Homebrew/test/support/fixtures/testball_bottle.rb [hunk 13]
```diff
@@ -6,7 +6,7 @@ def initialize(name = "testball_bottle", path = Pathname.new(__FILE__).expand_pa
       stable.bottle do
         cellar :any_skip_relocation
         root_url "file://#{TEST_FIXTURE_DIR}/bottles"
-        sha256 "9abc8ce779067e26556002c4ca6b9427b9874d25f0cafa7028e05b5c5c410cb4" => Utils::Bottles.tag
+        sha256 "d48bbbe583dcfbfa608579724fc6f0328b3cd316935c6ea22f134610aaf2952f" => Utils::Bottles.tag
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/fixtures/testball_bottle.rb]: Presumably: adding the formula?

File: Library/Homebrew/formula_installer.rb (hunk 11 above)
Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: `formula.bottle` returns `nil` when `compatible_cellar?` is false.
See https://github.com/Homebrew/brew/blob/master/Library/Homebrew/formula.rb#L330
and https://github.com/Homebrew/brew/blob/master/Library/Homebrew/software_spec.rb#L93

Rather than `formula.bottle_specification.compatible_cellar?` we could use `formula.bottled?`. Do you have a preference?

File: Library/Homebrew/test/support/fixtures/testball_bottle.rb (hunk 13 above)
Comment (sjackman) [reply Library/Homebrew/test/support/fixtures/testball_bottle.rb]: Yep

File: Library/Homebrew/formula_installer.rb (hunk 12 above)
Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: Ah, got it. Sorry I misunderstood. I've updated the warning.

File: Library/Homebrew/formula_installer.rb (hunk 11 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formula_installer.rb]: `formula.bottled?` would be preferable, yeh.

File: Library/Homebrew/formula_installer.rb [hunk 14]
```diff
@@ -236,6 +236,15 @@ def install
       raise CannotInstallFormulaError, message
     end
 
+    # Warn if a more recent version of this formula is available in the tap.
+    begin
+      if formula.pkg_version < (v = Formulary.factory(formula.full_name).pkg_version)
+        opoo "#{formula.name} #{v} is available and more recent than version #{formula.pkg_version}."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Use `formula.full_name` here? @ilovezfs, does this look OK?

Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: Done.

File: Library/Homebrew/formula_installer.rb (hunk 11 above)
Comment (sjackman) [reply Library/Homebrew/formula_installer.rb]: Done.

**Comment by MikeMcQuaid** [comment]
Thanks again @sjackman!

**Comment by sjackman** [comment]
Woo hoo! Thanks for merging, Mike!
